admin-dev/themes/new-theme/js/pages/product/components/categories/category-tree-selector.ts

```diff
@@ -0,0 +1,379 @@
+      const checkboxInput = inputsContainer.querySelector(ProductCategoryMap.treeCheckboxInput) as HTMLInputElement;
+      checkboxInput.value = String(treeCategory.id);
+
+      inputsContainer.addEventListener('click', (event) => {
+        // We don't want to mess with the inputs behaviour (no toggle when checkbox or radio is clicked)
+        // So we only toggle when the div itself is clicked.
+        if (event.target !== event.currentTarget) {
+          return;
+        }
+
+        const isExpanded = !childrenList.classList.contains('d-none');
+        categoryNode.classList.toggle('less', !isExpanded);
+        categoryNode.classList.toggle('more', isExpanded);
+        childrenList.classList.toggle('d-none', isExpanded);
+      });
+
+      // Recursively build the children trees
+      treeCategory.children.forEach((childCategory) => {
+        const childTree = this.generateCategoryTree(childCategory);
+
+        childrenList.append(childTree);
+      });
+    }
+
+    return categoryNode;
+  }
+
+  private generateTreeElement(category: TreeCategory): HTMLElement {
+    const categoryTree = this.categoryTree as HTMLElement;
+    const prototypeTemplate = categoryTree.dataset.prototype as string;
+    const prototypeName = categoryTree.dataset.prototypeName as string;
+
+    // It is convenient to have categoryId as prototype name index because it is unique, but nothing depends on it
+    const template = prototypeTemplate.replace(new RegExp(prototypeName, 'g'), String(category.id));
+    // Trim is important here or the first child could be some text (whitespace, or \n)
+    const frag = document.createRange().createContextualFragment(template.trim());
+    const categoryNode = frag.firstChild as HTMLElement;
+
+    // Add category name text
+    const checkboxInput = categoryNode.querySelector(ProductCategoryMap.checkboxInput) as HTMLInputElement;
+    checkboxInput.value = String(category.id);
+
+    const nameElement = document.createTextNode(category.name);
+    const element = category.active
+      ? nameElement
+      : document.createElement('i').appendChild(nameElement).parentNode as HTMLElement;
+
+    (checkboxInput.parentNode as HTMLElement).insertBefore(element, checkboxInput);
+
+    return categoryNode;
+  }
+
+  /**
+   * Check/uncheck the selected category (matched by its ID) and toggle the tree by going up through the category's ancestors.
+   */
+  private updateCategory(categoryId: number, check: boolean): void {
+    const treeElement = this.categoryTree as HTMLElement;
+    const checkbox = treeElement.querySelector(ProductCategoryMap.inputByValue(categoryId)) as HTMLInputElement;
+    checkbox.checked = check;
+    this.openCategoryParents(checkbox);
+    this.openCategoryParents(checkbox);
+    this.updateSelectedCategories();
+  }
+
+  private openCategoryParents(checkbox: HTMLInputElement): void {
```

Comment:
As discussed in standup, Ill leave it now as it is, because this little "feature" seems to be changing whenever I ask about it :smile: So we can adjust the behavior when we are certain about it in other PR's